### PR TITLE
SPARK-13052 waitingApps metric doesn't show the number of apps currently in the WAITING state

### PR DESCRIPTION
**Pull request by RaafatAkkad**

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala [hunk 1]
```diff
@@ -42,6 +42,6 @@ private[spark] class MasterSource(val master: Master) extends Source {
 
   // Gauge for waiting application numbers in cluster
   metricRegistry.register(MetricRegistry.name("waitingApps"), new Gauge[Int] {
-    override def getValue: Int = master.waitingApps.size
+    override def getValue: Int = master.apps.filter(_.state == ApplicationState.WAITING).size
```
Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala]: Do you always see these two values are different, from my understanding they should be same, perhaps there's some race condition which makes it different.

Comment (RaafatAkkad) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala]: @jerryshao In all my tests using multiple Spark versions, the "apps" and "waitingApps" metrics have been the same. Do you know if the "waitingApps" metric is intended to refer to something other than the number of apps that are currently in the waiting state?
Thanks

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala]: "waitingApps" suppose to refer to the number of waiting applications in the Master, and `apps` includes all the uncompleted applications including `waitingApps`. If there's any inconsistent meaning, I think we should find out what actual `waitingApps` usage in `Master`, to see if there's any bug in it.

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala]: Digging a bit about current usage of `waitingApps`, I think it cannot be well reflect the actual meaning of waiting applications, even RUNNING applications are still in the `waitingApps` list. So I think your fix is simple and effective.

**Comment by andrewor14** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/master/MasterSource.scala]: ok, I agree

**Comment by andrewor14** [comment]
@RaafatAkkad looks like `Master#waitingApps` no longer needs to be exposed. Could you make it a `private val`?

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
LGTM

**Comment by RaafatAkkad** [comment]
Thanks for the feedback. The tests have passed. Can you please re-trigger a CI build for the "Make waitingApps private" commit as it has failed?

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
Actually the tests passed but not for the latest commit. Let's re-run this.

**Comment by andrewor14** [comment]
Merged into master.

**Pull request closed without merging**